src/libopensc/card-piv.c

```diff
@@ -2440,15 +2440,16 @@ static int piv_select_file(sc_card_t *card, const sc_path_t *in_path,
 	pathlen = in_path->len;
 
 	/* only support single EF in current application */
+	/* 
+	 * PIV emulates files, and only does so becauses sc_pkcs15_* uses
+	 * select_file and read_binary. The emulation adds path emulated structures
+	 * so piv_select_file will find it.
+	 * there is no dir. Only direct access to emulated files
+	 * thus opensc-tool and opensc-explorer can not read the emulated files
+	 */
 
 	if (memcmp(path, "\x3F\x00", 2) == 0) {
-		if (pathlen == 2)   {
-			r = piv_select_aid(card, piv_aids[0].value, piv_aids[0].len_short, NULL, NULL);
-			SC_TEST_RET(card->ctx, SC_LOG_DEBUG_NORMAL, r, "Cannot select PIV AID");
-
-			SC_FUNC_RETURN(card->ctx, SC_LOG_DEBUG_NORMAL, r);
-		}
-		else if (pathlen > 2) {
+		if (pathlen > 2) {
```

Comment:
The previous code should not have tried to catch the MF. It should return an error. I do not want to keep the previous
semantics.

The card-piv.c emulates the select_file and read_binary and only works with paths defined in
card-piv.c as static const struct piv_object piv_objects[] to map a path to an object on the card
Get DATA is then used to read the object.

pkcs15-piv.c also has these paths. There are set when creating the emulated pkcs15 structures.
It may add CECE for certificates, as the certificate is contained  in a certificate object,
(along with other flags, and the certificate my be zipped to.) The pkcs15_pubkey is obtained
from the certificate. pkcs15_privkey type, size and parameters are obtained from the public key.

The values used for these paths are defined in NIST 800-73-3, as a hold over from previous federal smart cards,
but there is no requirement for the card to support access using select and read_binary. GET_DATA
is used instead.

The above tries to stick to the NIST 800-73-3 standards as close as possible.
There are also some nonstandard objects in card-piv.c to represent keys:
  295 /* following not standard , to be used by piv-tool only for testing */


*BUT* in the future if the standards change, or there is some compelling reason to extend the functionality
this code could be extended.



On 10/16/2015 12:36 PM, Frank Morgner wrote:
> In src/libopensc/card-piv.c <https://github.com/OpenSC/OpenSC/pull/579#discussion_r42268601>:
>
>>
>>  	if (memcmp(path, "\x3F\x00", 2) == 0) {
>> -		if (pathlen == 2)   {
>> -			r = piv_select_aid(card, piv_aids[0].value, piv_aids[0].len_short, NULL, NULL);
>> -			SC_TEST_RET(card->ctx, SC_LOG_DEBUG_NORMAL, r, "Cannot select PIV AID");
>> -
>> -			SC_FUNC_RETURN(card->ctx, SC_LOG_DEBUG_NORMAL, r);
>> -		}
>> -		else if (pathlen > 2) {
>> +		if (pathlen > 2) {
>
> /If/ you want to return an error if MF is selected, then your changes are totally OK. I just remarked that the code previously tried to catch MF while /not/ returning an error. This is a small change
> in behavior that may have unexpected results.
>
> If you want to fix the NULL dereference but keep the current semantics, you simply need to handle the |file_out| object.
>
> —
> Reply to this email directly or view it on GitHub <https://github.com/OpenSC/OpenSC/pull/579/files#r42268601>.
>

-- 

  Douglas E. Engert  <DEEngert@gmail.com>